### ISSUE TITLE
"--user" broken on Windows?

### ISSUE THREAD

**Issue by larsxschneider** [bagder/curl#573]
Curl auth hangs on Windows in "git-bash.exe" from the latest Git for Windows distribution (2.6.4):
```
curl --user myuser https://myserver
```
It works on Windows if I use "git-cmd.exe" and on OSX.

```
$ curl --version
curl 7.45.0 (i686-w64-mingw32) libcurl/7.45.0 OpenSSL/1.0.2d zlib/1.2.8 libidn/1.32 libssh2/1.6.0 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smtp smtps telnet tftp
Features: IDN IPv6 Largefile SSPI Kerberos SPNEGO NTLM SSL libz TLS-SRP
```

**Comment by larsxschneider** [comment]
Might be interesting for @dscho , too.

**Comment by bagder** [comment]
Please elaborate on why this is a curl problem and not a git problem?

**Comment by larsxschneider** [comment]
Oops... to much copy and paste. I fixed the command in the issue. I am invoking curl and therefore I think it is a curl problem. Is there a way for me to provide more debug information?

**Comment by bagder** [comment]
--user with only a user name and no password prompts for a password for me:
~~~~
$ curl --user hello localhost
Enter host password for user 'hello':

**Comment by gvanem** [comment]
You mean:
```
curl --user myuser https://myserver
```

is *hanging* waiting for a password? You can always put the login information in `~./.netrc` (or `~./_netrc` on Windows) to get logged in automatically.

**BTW**. IMHO it would be nice if curl's `_netrc` parses on Windows could accept a line like:
 ```
-include %APPDATA%/.netrc
 ```

Since it is more work to maintain 2 `*netrc` files. :+1: or :-1:  ?

**Comment by gvanem** [comment]
> $ curl --user hello localhost
> Enter host password for user 'hello':

Should not:
```
 echo secret  | curl --user hello --pass -  localhost
```

also work on a non-Windows platform? It doesn't on Windows due to it's use of `getch()` in `src/tool_getpass.c`. Replacing `getch()` with `getchar()`, the `echo |`form works on Windows too.

**Comment by larsxschneider** [comment]
Well, my goal is to see the prompt ... and this is not showing up. Any idea how I can debug this?

**Comment by larsxschneider** [comment]
@gvanem No, the call is not "hanging waiting for a password". I can't even enter the password. My intention here is not automation. I want the user to enter the password.

**Comment by gvanem** [comment]
You wrote:
> Curl auth hangs on Windows in "git-bash.exe" from the latest Git for Windows distribution (2.6.4):

I use [this Git-distro](https://git-for-windows.github.io/) myself. The curl.exe bundled with it works fine here. If I do a:
```
%ProgramFiles%\Git-2\mingw64\bin\curl.exe --user hello curl.haxx.se
```
I can see it is waiting inside `getch()`. From *Process Explorer*:
```
ntoskrnl.exe!KeSynchronizeExecution+0x3de6
ntoskrnl.exe!KeWaitForMultipleObjects+0x1320
ntoskrnl.exe!KeWaitForMultipleObjects+0xd38
ntoskrnl.exe!KeWaitForMutexObject+0x385
ntoskrnl.exe!NtDeviceIoControlFile+0xca1
ntoskrnl.exe!NtDeviceIoControlFile+0x56
ntoskrnl.exe!setjmpex+0x3823
ntdll.dll!NtDeviceIoControlFile+0xa
KERNELBASE.dll!SetConsoleMode+0x16e
KERNELBASE.dll!PeekConsoleInputW+0xf1
KERNELBASE.dll!ReadConsoleInputA+0x14
msvcrt.dll!getch+0xcb
msvcrt.dll!getch+0x16
curl.exe+0x7de3
```
I'm pretty sure this is a problem related to the *bash* used in that git-bash.

**Comment by bagder** [comment]
@larsxschneider, since it works for @gvanem and the windows function in curl for reading passwords (https://github.com/bagder/curl/blob/master/src/tool_getpass.c#L105) hasn't been changed since 2011 in any substantial way, I too lean towards pointing the blame for this on something else.

**Comment by dscho** [comment]
> I too lean towards pointing the blame for this on something else

@bagder and you would be correct. The problem is the terminal emulator we use with Git Bash since Git for Windows 2.5, MinTTY. This terminal emulator is *not* associated with a Win32 Console, therefore the user does not see anything when cURL wants to interact with the user via said Console.

I have not found a way (yet) to have a Console in the waiting for such issues, but I do have a work around (mentioned in Git for Windows' [known issues](https://github.com/git-for-windows/build-extra/blob/master/installer/ReleaseNotes.md#known-issues):

> * Some console programs interact correctly with MinTTY only when called through `winpty` (e.g. the Python console needs to be started as `winpty python` instead of just `python`).

**Comment by gvanem** [comment]
> This terminal emulator is not associated with a Win32 Console, 

Since it is a GUI-program, that's no surprise (maybe there is a `mintty` linked as a *console* app. That would make a difference).

But why do you need to use Bash interactively (like from inside `git-bash`)? I *always* use `git`-program from 4NT (or cmd) and it works just fine.

**Comment by dscho** [comment]
> > This terminal emulator is not associated with a Win32 Console,
> 
> Since it is a GUI-program, that's no surprise (maybe there is a mintty linked as a console app. That would make a difference).

Actually, there are examples of terminal emulators that *are* GUI programs, yet they still bring their own Win32 Console just so that programs like cURL can interact with that Console. The ugly part is that they have to poll the (hidden) Console in order to display the text.

> But why do you need to use Bash interactively (like from inside git-bash)? I always use git-program from 4NT (or cmd) and it works just fine.

![dude-thats-just-like-your-opinion-man_zps9268c8a7](https://cloud.githubusercontent.com/assets/127790/12053170/b99c91c0-af16-11e5-840a-34f1667aa744.jpg)

Sorry, could not resist... :stuck_out_tongue_winking_eye:

**Comment by MoSal** [comment]
@dscho 
I've got [confused about mintty not being a windows console myself](https://github.com/mintty/mintty/issues/482).

mintty uses Cygwin or MSYS(2) runtimes. You need a curl binary that is built and linked against the runtime your git distribution is using.

mingw is used to build *native* binaries. That's why it's not working for you.

**Comment by dscho** [comment]
> mingw is used to build native binaries.

Correct.

In fact, Git for Windows' *Git* executables *are* native Win32 binaries built into MinGW package.

> That's why it's not working for you.

That is not quite correct. We actually made things work for the interaction between `git.exe` and MinTTY's console, but I do agree that this is a bit out of scope for cURL. In any case, Git for Windows *also* ships the MSys2 `curl.exe` which is able to talk to MinTTY's (Cygwin's) console, so @larsxschneider could use `/usr/bin/curl.exe` *or* `winpty /mingw64/bin/curl.exe` to work around this quirk.

Short version: I am half-convinced that this ticket should be closed as "Won't Fix" because it would require code too specific to MSys2 to work around.

**Comment by jay** [comment]
> Short version: I am half-convinced that this ticket should be closed as "Won't Fix" because it would require code too specific to MSys2 to work around.

@dscho I agree but I'd like some understanding of how you did it. From what I know ReadConsoleInput requires a console and won't return until it has some input. The alternative would be do something other than getch to read from what I assume is a pipe.

Related: I've had problems with stdio/stderr buffering when running native apps in Cygwin mintty. The stdio and stderr can be jumbled together. It may not be flushed the way one would expect. I'm not sure if this is true of all mintty.

Also see mintty issue [Improve support for native console programs](https://code.google.com/p/mintty/issues/detail?id=56) (since migrated to mintty/mintty#56).

**Comment by dscho** [comment]
> I'd like some understanding of how you did it. From what I know ReadConsoleInput requires a console and won't return until it has some input. The alternative would be do something other than getch to read from what I assume is a pipe.

Here you are: https://github.com/git-for-windows/git/blob/27c08ea187462e56ffb514c8c997df419f004ce5/compat/winansi.c#L530-L569 (This is the part where we detect that a given file descriptor is connected to what MSys2/MinTTY regard a console.)

> I've had problems with stdio/stderr buffering when running native apps in Cygwin mintty.

Yes, this is a known issue. At some stage, I almost pulled my hair out while debugging a problem using `fprintf` statements because `stderr` was *buffered* while all my previous life I was used to it being unbuffered.

> I'm not sure if this is true of all mintty.

It is.

> Also see mintty issue Improve support for native console programs

Yeah, this is a problem that I underestimated. An additional complication is that I personally have a hard time to work with the mintty maintainer, so I fear I cannot help with this ticket...

**Comment by larsxschneider** [comment]
@dscho Thanks for the explanation and the work around :+1:

**Comment by dscho** [comment]
@larsxschneider thanks for caring and opening this ticket for future reference!

**Comment by bagder** [comment]
Thanks everyone. I see no remaining issue here for us to fix in curl so I'm closing it up now.

**Comment by dashesy** [comment]
So that answer is to use `winpty curl ...`, thanks
